src/avl_tree.py

```diff
@@ -0,0 +1,172 @@
+from __future__ import annotations
+from typing import Union
+
+
+class AVLTree:
+    class Node:
+        def __init__(
+            self,
+            val: float,
+            left: AVLTree.Node = None,
+            right: AVLTree.Node = None,
+        ):
+            self.val = val
+            self.left = left
+            self.right = right
+            self.height: int = 1
+
+        def __repr__(self):
+            left = 'na'
+            right = 'na'
+            if self.left is not None:
+                left = self.left.val
+            if self.right is not None:
+                right = self.right.val
+            return (
+                f"  {self.val}  \n"
+                f" /  \\  \n"
+                f"{left}   {right}")
+
+    def __init__(self, val: float):
+        self.head = self.Node(val)
+
+    def insert(self, val: float):
+        self.head = self.__insert(self.head, val)
+
+    def remove(self, val: float):
+        self.head = self.__remove(self.head, val)
+
+    def __insert(self, root: AVLTree.Node, val: float):
+        """
+        Insert a new value into the tree while maintain the AVL structure.
+        Insertion should takes O(logN) in time complexity.
+        Needs recursion to update height of each node.
+        """
+        if root is None:
+            return self.Node(val)
+        if val < root.val:
+            root.left = self.__insert(root.left, val)
+        elif val > root.val:
+            root.right = self.__insert(root.right, val)
+        else:
+            raise ValueError(f"Value {val} already exists!")
+
+        root.height = (
+            1 + 
+            max(
+                self.__get_height(root.left),
+                self.__get_height(root.right)))
```

Comment:
IMO, we should create a method for this part and reuse to avoid duplicating code. For example `__update_height(root)`